### ISSUE TITLE
More convenient defaults / null safety for nested structures

### ISSUE THREAD

**Issue by Gama11** [elnabo/json2object#19]
`@:default()` is a great way to avoid null-checks and allows for maximum granularity. However, with nested structures, it also get cumbersome / redundant quickly.

This works great:

```haxe
class Main {
    public static function main() {
        var parser = new json2object.JsonParser<Struct>();
        trace(parser.fromJson('{ "s": {} }', ""));
    }
}

typedef Struct = {
    var s:InnerStruct;
}

typedef InnerStruct = {
    @:optional @:default(0) var a:Int;
    @:optional @:default(0) var b:Int;
    @:optional @:default(0) var c:Int;
    @:optional @:default(0) var d:Int;
}
```

However, imagine we make `s` optional itself, so instead of parsing `'{ "s": {} }'`, we want to parse just `'{}'`. We can't just use `@:default({})` for it, because then we end up with null fields in the inner struct, despite their defaults. We have to specify everything redundantly: `@:default({a: 0, b: 0, c: 0, d: 0})`

The meaning of `@:default({})` shouldn't be changed either, because then you would lose the ability to express that.

Maybe this can be supported via some special syntax, like `@:default(auto)` (feel free to use a better name for it :D)? `auto` would use `0` for Int / Float, `false` for Bools, `[]` for arrays, and properly fill out the fields with their respective fields for structures.

Along those lines, it would also be nice if a structure could be parsed in a null-safe way _without having to add a meta to every single one of the fields_. Perhaps there could be an option for this in `fromJson()`, that just uses the `@:default(auto)` behavior for everything automatically?

**Comment by elnabo** [comment]
At the moment, even without a `@:default` meta `Int`, `Float` and `Bool` types are defaulted to `0`, `0.0` and `false`.

As for `auto` to initialize structures, this seems like an interesting idea, but recursive structures would be problematic. Paging @ibilon

**Comment by Gama11** [comment]
>At the moment, even without a @:default meta Int, Float and Bool types are defaulted to 0, 0.0 and false.

Are you sure? This traces `{ s: { a: null } }`.

```haxe
class Main {
    public static function main() {
        var parser = new json2object.JsonParser<Struct>();
        trace(parser.fromJson('{"s": {}}', ""));
    }
}

typedef Struct = {
    @:optional var s:InnerStruct;
}

typedef InnerStruct = {
    @:optional var a:Int;
}
```

I guess that's because `a` is `@:optional`, and thus the type becomes `Null<Int>`. An oversight?

**Comment by ibilon** [comment]
The float/int/bool are only automatically defaulted for structures (since we do new's job manually).

Maybe `@:default(notnull)` for value, not sure.
And good point about the default default (:p) behavior.

**Comment by elnabo** [comment]
If I read it correctly this must be a bug, as Int in anonymous structure are supposed to be initialized either as `0` or at the `@:default` value.

**Comment by ibilon** [comment]
> I guess that's because a is @:optional, and thus the type becomes Null\<Int>. An oversight?

Possible yes.

**Comment by elnabo** [comment]
The null value for Int/Float/Bool without default should be fixed by a2046c9be0e318b53d1bcb574b3bc8d098171c14

**Comment by ibilon** [comment]
Support for `@:default(auto)` was added in https://github.com/elnabo/json2object/commit/ac251e1d862e519661604efc64461278755f9a83
It'll make a structure/object with default value for its field instead of setting to null if it's not in the json.
